### PR TITLE
sns_topic - Add tags and purge_tags options

### PR DESCRIPTION
**Pull request by alinabuzachis**
##### SUMMARY

sns_topic - Add tags and purge_tags options

Closes https://github.com/ansible-collections/community.aws/issues/964

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

sns_topic

### CODE REVIEW

**Review by marknet15 — APPROVED** [review]
Nice 👍

File: plugins/modules/sns_topic.py [hunk 1]
```diff
@@ -462,6 +470,10 @@ def ensure_ok(self):
         elif self.display_name or self.policy or self.delivery_policy:
             self.module.fail_json(msg="Cannot set display name, policy or delivery policy for SNS topics not owned by this account")
         changed |= self._set_topic_subs()
+        
+        # Check tagging
+        changed |= update_tags(self.connection, self.module, self.topic_arn)
```
Comment (marknet15) [inline plugins/modules/sns_topic.py:475]: Not looked fully into it but the tests seem to fail on this part

**Comment by alinabuzachis** [comment]
recheck

**Comment by marknet15** [comment]
I've had a look at the tests failure and what I can see is:
```
botocore.errorfactory.AuthorizationErrorException: An error occurred (AuthorizationError) when calling the ListTagsForResource operation: User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=remote=zuul-cloud is not authorized to perform: SNS:ListTagsForResource on resource: arn:aws:sns:us-east-1:966509639900:ansible-test-31718258-node-0002282009-topic because no identity-based policy allows the SNS:ListTagsForResource action
```

So looks like its just a case of missing permissions

File: plugins/modules/sns_topic.py [hunk 2]
```diff
@@ -140,6 +140,16 @@
...
+    type: dict
+    version_added: 3.2.0
+  purge_tags:
+    description: Set to C(False) to retain any tags that aren't specified in task and are associated with the instance.
+    type: bool
+    default: true
+    version_added: 3.2.0
 extends_documentation_fragment:
 - amazon.aws.aws
 - amazon.aws.ec2
```
Comment (tremble) [inline plugins/modules/sns_topic.py:155]:
```suggestion
notes:
- Support for I(tags) and I(purge_tags) was added in release 4.0.0.
extends_documentation_fragment:
- amazon.aws.aws
- amazon.aws.ec2
- amazon.aws.tags
```

Comment (alinabuzachis) [reply plugins/modules/sns_topic.py]: Done, thanks!

File: plugins/module_utils/sns.py [hunk 3]
```diff
@@ -87,6 +91,13 @@ def canonicalize_endpoint(protocol, endpoint):
     return endpoint
 
 
+def get_tags(client, module, topic_arn):
+    try:
+        return boto3_tag_list_to_ansible_dict(client.list_tags_for_resource(ResourceArn=topic_arn)['Tags'])
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
```
Comment (tremble) [inline plugins/module_utils/sns.py:97]:
```suggestion
    except is_boto3_error_code('AuthorizationError'):
        module.warn('Unable access tags resource {0}'.format(topic_arn)
    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
```

**Pull request merged**